src/jit/decomposelongs.cpp

```diff
@@ -494,32 +494,100 @@ GenTree* DecomposeLongs::DecomposeCast(LIR::Use& use)
+    var_types dstType = cast->CastToType();
+
+    if ((cast->gtFlags & GTF_UNSIGNED) != 0)
+    {
+        srcType = genUnsignedType(srcType);
+    }
+
+    if (varTypeIsLong(srcType))
+    {
+        if (cast->gtOverflow() && (varTypeIsUnsigned(srcType) != varTypeIsUnsigned(dstType)))
+        {
+            GenTree* srcOp = cast->gtGetOp1();
+            noway_assert(srcOp->OperGet() == GT_LONG);
+            GenTree* loSrcOp = srcOp->gtGetOp1();
+            GenTree* hiSrcOp = srcOp->gtGetOp2();
+
+            //
+            // When casting between long types an overflow check is needed only if the types
+            // have different signedness. In both cases (long->ulong and ulong->long) we only
+            // need to check if the high part is negative or not. Use the existing cast node
+            // to perform a int->uint cast of the high part to take advantage of the overflow
+            // check provided by codegen.
+            //
+
+            loResult = loSrcOp;
+
+            hiResult                       = cast;
+            hiResult->gtType               = TYP_INT;
+            hiResult->AsCast()->gtCastType = TYP_UINT;
+            hiResult->gtFlags &= ~GTF_UNSIGNED;
+            hiResult->gtOp.gtOp1 = hiSrcOp;
+
+            Range().Remove(cast);
+            Range().Remove(srcOp);
+            Range().InsertAfter(hiSrcOp, hiResult);
+        }
+        else
+        {
+            NYI("Unimplemented long->long no-op cast decomposition");
+        }
+    }
+    else if (varTypeIsIntegralOrI(srcType))
     {
-        case TYP_INT:
-            if (tree->gtFlags & GTF_UNSIGNED)
+        if (varTypeIsUnsigned(srcType))
+        {
+            loResult = cast->gtGetOp1();
+            hiResult = m_compiler->gtNewZeroConNode(TYP_INT);
+
+            Range().Remove(cast);
+            Range().InsertAfter(loResult, hiResult);
+        }
+        else
+        {
+            if (cast->gtOverflow() && varTypeIsUnsigned(dstType))
             {
-                loResult = tree->gtGetOp1();
-                Range().Remove(tree);
+                //
+                // An overflow check is needed only when casting from a signed type to ulong.
+                // Change the cast type to uint to take advantage of the overflow check provided
+                // by codegen and then zero extend the resulting uint to ulong.
+                //
+
+                loResult                       = cast;
+                loResult->AsCast()->gtCastType = TYP_UINT;
+                loResult->gtType               = TYP_INT;
+
+                hiResult = m_compiler->gtNewZeroConNode(TYP_INT);
 
-                hiResult = new (m_compiler, GT_CNS_INT) GenTreeIntCon(TYP_INT, 0);
                 Range().InsertAfter(loResult, hiResult);
             }
             else
             {
-                NYI("Lowering of signed cast TYP_INT->TYP_LONG");
-            }
-            break;
+                LIR::Use src(Range(), &(cast->gtOp.gtOp1), cast);
```

Comment:
@pgavlin Rather than adding 2 new comments I reorganized the code so that the overflow case is handled first instead of being intermixed with the other 2 cases. Together with an already existing comment I think it should be clear enough.